common4j/src/main/com/microsoft/identity/common/java/cache/MicrosoftStsAccountCredentialAdapter.java

```diff
@@ -268,6 +284,195 @@ public IdTokenRecord asIdToken(MicrosoftAccount msAccount, MicrosoftRefreshToken
         return idToken;
     }
 
+    @Override
+    public AccountRecord createAccountRecord(
+            @NonNull final TokenCommandParameters parameters,
+            final SdkType sdkType,
+            @NonNull final MicrosoftStsTokenResponse microsoftStsTokenResponse) throws ServiceException {
+        final String methodTag = TAG + ":createAccountRecord";
+        final ClientInfo clientInfo = new ClientInfo(microsoftStsTokenResponse.getClientInfo());
+        AccountRecord accountRecord;
+
+        if (sdkType.equals(SdkType.ADAL)) {
+            final AzureActiveDirectoryAccount azureActiveDirectoryAccount =
+                    new AzureActiveDirectoryAccount(
+                            new IDToken(microsoftStsTokenResponse.getIdToken()),
+                            clientInfo
+                    );
+            accountRecord = new AccountRecord(azureActiveDirectoryAccount);
+        } else {
+            final MicrosoftStsAccount microsoftStsAccount = new MicrosoftStsAccount(
+                    new IDToken(microsoftStsTokenResponse.getIdToken()),
+                    clientInfo
+            );
+            accountRecord = new AccountRecord(microsoftStsAccount);
+        }
+
+        URL authority = parameters.getAuthority().getAuthorityURL();
+
+        // If token response has authority set it else fallback to request authority
+        if (!microsoftStsTokenResponse.getAuthority().isEmpty()) {
+            try {
+                authority = new URL(microsoftStsTokenResponse.getAuthority());
+            } catch (MalformedURLException e) {
+                Logger.error(TAG,
+                        "Authority url construction failed, setting request authority to result",
+                        e
+                );
+            }
+        }
+
+        final AzureActiveDirectoryCloud cloudEnv =
+                AzureActiveDirectory.getAzureActiveDirectoryCloud(authority);
+
+        if (cloudEnv != null) {
+            Logger.info(methodTag, "Using preferred cache host name...");
+            accountRecord.setEnvironment(cloudEnv.getPreferredCacheHostName());
+        } else {
+            accountRecord.setEnvironment(authority.getHost());
```

Comment:
is it possible to have authority as null here? I see that is being assigned above from parameters, but want to check if it can be null at this point?

Reply:
I don't think so. At least TokenCommandParameters always has it.